8강 확증적 데이터 분석 CDA
통계적 가설검정 가설을 수립해 객관적인 수치로 가설의 참/거짓을 규명
귀무가설:0이다
집단간 평균이 같다/정규분포/상관성이 없다/연관성이 없다

대립가설:0이 아니다
정규분포x 집단간 평균이 다르다/ 상관성이 있다
연관성이 있는 상황

통계학에서 기본적인 가정:항목간에 독립된 상태로 데이터가 수집

p-value 귀무가설 참 지지도

유의 수준(5%):수립된 가설을 채틱/기각  시키기 위한 기준
P-value >5% 귀무가설 참
P-value <5% 귀무가설 거짓

단일변수:
1)연속형(숫자):해당 숫자데이터의 분포가 정규분포를 따르는가? (정규성검정Normal Test)

->다른 통계적인 분석을 수행할 때, 숫자데이터의 분포의 모양에 따라 분석 기법이 달라짐

->귀무가설:해당 데이터 분포는 정규분포를 따른다

->대립가설:해당 데이터 분포는 정규분포를 따르지 않는다

다변수:
1)x변주형/Y연속형:집단 간 통계량 (평균/분산)을 비교

2)x연속형/Y연속형:두연속형 숫자데이터의 상관성을 확인

3)x범주형/Y범주형:두 범주형 데이터가 서로연관/독립인지 확인
->카에제곱검정



In [ ]:
import numpy as np
import pandas as pd
from google.colab import files
from google.colab import drive
drive.mount('/content/drive')
trian = pd.read_csv("/content/drive/MyDrive/DSA/03_Data.csv")
df1=pd.read_csv("/content/drive/MyDrive/DSA/03_Data.csv")




Mounted at /content/drive


#통계적 가설 검정

In [ ]:
#고객이 유입된 '채널'별 '최종결제금액'의 평균이 차이가 있는지 확인
df1.pivot_table(index='채널',values='최종결제금액',aggfunc='mean')

최종결제금액
채널                      
direct     105512.323232
email      113118.376384
facebook   119932.625000
google     105529.000000
instagram  107958.181818
naver      110497.274529
others     107445.820144

facebook으로 유입된 고객들과 instagram 으로 유입된 고객의 최종결제 금액의 평균의 차이가 있는가?

In [ ]:
cond1=(df1['채널']=='facebook')
cond2=(df1['채널']=='instagram')
df1.loc[cond1]
df1.loc[cond2]

회원코드              프로그램         상태  프로그램가격  최종결제금액  \
82    RDZYYEGX   체중감량 프로그램 (타입B)  cancelled  112500  112500   
309   CYKXD5S9  체중감량 프로그램 (타입 A)  completed  112500  112500   
502   2Y232WHI  체중감량 프로그램 (타입 A)  cancelled  112500  112500   
512   UYQYNM26  체중감량 프로그램 (타입 A)  completed  112500  112500   
1222  F91KD0N0   체중감량 프로그램 (타입B)  completed  112500  112500   
1295  5D3QTOWQ   체중감량 프로그램 (타입B)  completed  112500  112500   
1890  EYG0MLDU   체중감량 프로그램 (타입B)  cancelled   44780   44780   
3134  NT9R87HV  체중감량 프로그램 (타입 A)  completed  112500  112500   
3157  R6Q23VN6   체중감량 프로그램 (타입B)  cancelled  112500  112500   
3485  O2KBCUAZ  체중감량 프로그램 (타입 A)  completed  112500  112500   
3497  U3NR4G3L   체중감량 프로그램 (타입B)   refunded  112500  137500   
3501  3B4JOOJK   체중감량 프로그램 (타입B)  completed  112500  112500   
3651  HA4GU94Q  체중감량 프로그램 (타입 A)  cancelled  112500  112500   
4204  724MWDJ1   체중감량 프로그램 (타입B)   refunded  112500  137500   
4272  BGGYFFJL   체중감량 프로그램 (타입B)   refunded  112500  137500   
4687  MBEJQANA   체중감량 프로그램 (타입B)  completed  112500  112500   
4843  WLBG0XRQ  체중감량 프로그램 (타입 A)  completed  112500  112500   
4953  D9S7ZYLD  체중감량 프로그램 (타입 A)  completed  112500  112500   
6121  L3BM2PC7   체중감량 프로그램 (타입B)  cancelled  112500  112500   
6229  LRA6O0KQ  체중감량 프로그램 (타입 A)  cancelled  112500  112500   
6589  6EH2LGR5   체중감량 프로그램 (타입B)  completed  112500  137500   
6660  GV11KU81  체중감량 프로그램 (타입 A)  completed  112500  112500   
6716  W5TD4E9F   체중감량 프로그램 (타입B)  cancelled   44780   44780   
7143  Z4T57DFC   체중감량 프로그램 (타입B)  cancelled  112500  112500   
7551  2ECGJU4M   체중감량 프로그램 (타입B)  cancelled   44780   44780   
7671  U866NFE4   체중감량 프로그램 (타입B)  completed  112500  112500   
7781  OT5EU5LW  체중감량 프로그램 (타입 A)  cancelled  112500  112500   
7909  V2U4PFMO   체중감량 프로그램 (타입B)  completed   44780   44780   
8534  3JMENOIT  체중감량 프로그램 (타입 A)  completed  112500  112500   
8780  U2K07BJK  체중감량 프로그램 (타입 A)  completed  112500  112500   
8818  ZYBWK2Q9  체중감량 프로그램 (타입 A)  completed  112500  133500   
9278  WTUBFGKW   체중감량 프로그램 (타입B)  cancelled  112500  112500   
9988  3W9NJ7A5  체중감량 프로그램 (타입 A)  cancelled  112500  112500   

                               결제시각       결제유형         채널  보조식품가격 보조식품(추가구매)  
82     2017-05-12 오후 4:42:37.304000  Recurring  instagram       0       구매안함  
309   2017-06-27 오후 12:44:28.152000  Recurring  instagram       0       구매안함  
502    2017-07-01 오후 3:37:58.424000  Recurring  instagram       0       구매안함  
512    2017-07-01 오후 3:44:04.424000  Recurring  instagram       0       구매안함  
1222   2017-05-16 오전 1:20:43.808000  Recurring  instagram       0       구매안함  
1295  2017-05-07 오후 10:51:15.416000  Recurring  instagram       0       구매안함  
1890  2017-04-17 오전 11:59:05.736000  Recurring  instagram       0       구매안함  
3134   2017-07-16 오후 1:02:56.080000  Recurring  instagram       0       구매안함  
3157   2017-05-20 오후 6:27:30.136000  Recurring  instagram       0       구매안함  
3485   2017-07-19 오후 1:47:51.432000  Recurring  instagram       0       구매안함  
3497  2017-05-05 오전 10:51:02.952000  Recurring  instagram   25000     종합비타민A  
3501  2017-05-18 오후 10:56:52.904000  Recurring  instagram       0       구매안함  
3651  2017-07-16 오후 12:57:06.080000  Recurring  instagram       0       구매안함  
4204  2017-05-05 오전 10:50:59.952000  Recurring  instagram   25000     종합비타민A  
4272  2017-05-05 오전 10:58:05.952000  Recurring  instagram   25000     종합비타민A  
4687   2017-05-12 오후 4:40:37.304000  Recurring  instagram       0       구매안함  
4843   2017-07-01 오후 3:31:32.424000  Recurring  instagram       0       구매안함  
4953  2017-06-27 오후 12:48:45.152000  Recurring  instagram       0       구매안함  
6121  2017-05-07 오후 10:37:26.416000  Recurring  instagram       0       구매안함  
6229   2017-07-19 오후 2:02:49.432000  Recurring  instagram       0       구매안함  
6589  2017-05-05 오전 10:59:05.952000  Recurring  instagram   25000     종합비타민A  
6660   2017-06-06 오후 6:20:23.384000  Recurring  instagram       0       구매안함  
6716   2017-05-02 오후 8:22:03.800000  Recurring  in

In [ ]:
df1_face=df1.loc[cond1]
df1_insta=df1.loc[cond2]

In [ ]:
#가설 검정을 수행하는 Scipy 라이브러리 설치(아나콘다x)
!pip--trusted-host pypi.org--trusted-host files.pythinhost.org.install scipy

/bin/bash: line 1: pip--trusted-host: command not found


In [ ]:
import scipy.stats as stats #가설 검정을 함수의 집합체 수행하는 라이브러리

[링크 텍스트](https://drive.google.com/file/d/119yyfiS7PqCJe6ixDRg6hEzShnCpTcPB/view?usp=drive_link)

#1-1각집단의 연속형 자료의 정규성 검정
#귀무가설:해당 숫자데이터는 정규분포를 따른다
#대립가설:해당 숫자데이터는 정규분포를 안따른다

In [ ]:
stats.normaltest(df1_face['최종결제금액']) #가설 검정

NormaltestResult(statistic=2204.8632503534313, pvalue=0.0)

NormaltestResult(statistic=2204.8632503534313, pvalue=0.0)라고 답이 나오고 pvalue는 0이다 귀무가설 참 지지도

유의수준 5%보다 >0 pvalue이기에 기각

In [ ]:
stats.normaltest(df1_insta['최종결제금액']) #가설 검정

NormaltestResult(statistic=19.080080970100784, pvalue=7.191393615906394e-05)

NormaltestResult(statistic=19.080080970100784, pvalue=7.191393615906394e-05)라고 나옴 0.000000005인거임 엄청작은수 인거지
5%>0.000000005 유의수준 기각이므로 대립가설 참이고, 정규분포를 따르지 않는다

1-2 정규분포를 따르지 않는 두집단 ;최종결제금액의 평균이 차이기 있는가?


귀무가설:두 집단의 평균의 차이가 없다
대립가설:두 집단 평균의 차이는 있다.

In [ ]:
stats.ranksums(df1_face['최종결제금액'],df1_insta['최종결제금액'])

RanksumsResult(statistic=3.3310612893175584, pvalue=0.0008651555174077354)

stats.ranksums(df1_face['최종결제금액'],df1_insta['최종결제금액'])
0.5보다 작기에 유의수준보다 작기에 0.000005 귀무가설기각이됨

->대립가설 참이기에 두집단 평균 차이는 있다

In [ ]:
df1

회원코드              프로그램         상태  프로그램가격  최종결제금액  \
0     Y9RY2VSI   체중감량 프로그램 (타입B)  completed  112500  112500   
1     3GTN3S3B   체중감량 프로그램 (타입B)  completed   44780   44780   
2     6B0IG276  체중감량 프로그램 (타입 A)  completed  132000  149000   
3     EMGRU2MO  체중감량 프로그램 (타입 A)  completed  112500  112500   
4     1ELG96TX   체중감량 프로그램 (타입B)  completed   44780   44780   
...        ...               ...        ...     ...     ...   
9995  118AFCQ9   체중감량 프로그램 (타입B)  completed  112500  112500   
9996  VDE8FXV9  체중감량 프로그램 (타입 A)  completed  112500  129500   
9997  VWJ4NLZY  체중감량 프로그램 (타입 A)  completed  112500  129500   
9998  NJ2PR967   체중감량 프로그램 (타입B)  cancelled  112500  129500   
9999  KVX64CPM  체중감량 프로그램 (타입 A)  cancelled  112500  112500   

                               결제시각       결제유형        채널  보조식품가격 보조식품(추가구매)  
0      2017-04-14 오후 7:03:29.976000  Recurring    others       0       구매안함  
1      2017-05-23 오후 8:53:54.368000  Recurring    others       0       구매안함  
2     2017-08-23 오후 11:39:21.840000  Recurring  facebook   17000   식이요법보조제A  
3      2017-08-28 오후 8:18:22.824000  Recurring     naver       0       구매안함  
4      2017-05-07 오후 5:50:30.944000  Recurring  facebook       0       구매안함  
...                             ...        ...       ...     ...        ...  
9995  2017-04-18 오전 12:43:58.104000  Recurring  facebook       0       구매안함  
9996  2017-08-16 오전 10:03:10.136000  Recurring  facebook   17000   식이요법보조제A  
9997  2017-08-09 오전 11:47:55.408000  Recurring  facebook   17000   식이요법보조제A  
9998   2017-04-28 오전 7:05:19.776000  Recurring  facebook       0       구매안함  
9999  2017-06-02 오전 10:16:34.576000  Recurring    others       0       구매안함  

[10000 rows x 10 columns]

In [ ]:
df1['보조식품(추가구매)'].value_counts()

보조식품(추가구매)
구매안함        7520
식이요법보조제A    1582
식이요법보조제B     388
단백질보충제A      340
종합비타민A       108
단백질보충제B       62
Name: count, dtype: int64

In [ ]:
df1['상태'].value_counts()

상태
completed    5400
cancelled    4010
refunded      590
Name: count, dtype: int64

카이제곱검정 x범주형/y범주형일때
#보조식품의 종류에 따라 결제상태가 달라지는가?
#보조식품(추가구매)와 '상태'가 서로 연관/독립인가?

In [ ]:
#crosstab(():두 범주형 항목에 대한 빈도수를 계산

pd.crosstab(df1['보조식품(추가구매)'],df1['상태'])

상태          cancelled  completed  refunded
보조식품(추가구매)                                
구매안함             4010       3179       331
단백질보충제A             0        302        38
단백질보충제B             0         53         9
식이요법보조제A            0       1412       170
식이요법보조제B            0        360        28
종합비타민A              0         94        14

In [ ]:
#상태 '항목'에서 canelled데이터를 제외한 고객들만 확인

In [ ]:
cond1=(df1['상태']!='Cancelled')
df2=df1.loc[cond1]

In [ ]:
df2_cross=pd.crosstab(df2['보조식품(추가구매)'],df2['상태'])

In [ ]:
stats.chi2_contingency(df2_cross)
#p.value 0.1294>0.05 5$ 귀무가설 참
#두데이터는 서로 독립이다
#보조식품의 종류에 따라 결재상태가 달라지지 않는다 연관성이 없다


Chi2ContingencyResult(statistic=2221.37801009406, pvalue=0.0, dof=10, expected_freq=array([[3.01552e+03, 4.06080e+03, 4.43680e+02],
       [1.36340e+02, 1.83600e+02, 2.00600e+01],
       [2.48620e+01, 3.34800e+01, 3.65800e+00],
       [6.34382e+02, 8.54280e+02, 9.33380e+01],
       [1.55588e+02, 2.09520e+02, 2.28920e+01],
       [4.33080e+01, 5.83200e+01, 6.37200e+00]]))

#x연속성/Y연속형 (두 숫자가 서로 상관성이 있는가?)
#'보조식품'을 구매한 고객들 중, '보조식품가격'과 '최종결재금액;의 상관성이 있는지?

In [ ]:
cond1=(df1['보조식품(추가구매)']!='구매안함')
df3=df1.loc[cond1]

#보조식품을 구매한 고객들만 데이터 프레임 형태로 추출

숫자 숫자면 정규성 검정필요
#1-1 '보조식품가격;과 '최종결제 금액'데이터가 정규분포를 띄는지?(정규성검정)
#귀무가설:해당 숫자데이터의 분포 정규분포는 따른다
#대립가설:해당 숫자데이터의 분포 정규분포는 안따른다


In [ ]:
stats.normaltest(df3['보조식품가격']) #정규분포 띄는지 확인 p-value 0.00000005 대립가설 참 정규분포 안따름

NormaltestResult(statistic=539.119973010461, pvalue=8.542506518439196e-118)

In [ ]:
stats.normaltest(df3['최종결제금액']) #정규분포 0.0000005 대립가설 참 정규분포 안따림

NormaltestResult(statistic=1146.4934414641564, pvalue=1.1018244793842086e-249)

#1-2 '보조식품가격'과 '최종결재금액;의 상관 관계이 있는지?
#귀무가설:두 연속형 데이터는 서로 상관성이 없다

#대립가설:두 연송형 데이터는 서로 상관성이 있다

In [ ]:
stats.spearmanr(df3['보조식품가격'],df3['최종결제금액']) #spearmanr 0.000005 대립가설 참 ->데이터는 서로 상관관꼐가 있다

SignificanceResult(statistic=0.6188102640438136, pvalue=4.3725233196527346e-262)

9.예측적 데이터 분석 PDA
기계학습 머신러닝:컴퓨터가 데이터간 (y목표변수/x설명 변수) 규칙/수식을 통해 도출하는 작업

학습핵심:point:

1)학습능력:컴퓨터가 데이터로부터 적절한 수식을 도출하는 능력

2)일반화 능력:학습을 통해 수식(모델)이 새로운 데이터에 대해 잘 예측/대응 하는 능력

미리 꼼쳐놓고 검증함


-기계학습 핵심 3요소

1.데이터(교과서):학습 목적에 맞게 데에터를 깔끔하게 다듬어야한다.
->특성공학(feture Engineering)

2.알고리즘(선생님):학습에 있어 데이터로부터 수식/규칙을 도출할 수 있는 적절한 알고리즘을 선택하는 것이 중요

선형회귀/의사결정나무/앙상블/신경망

3.하드웨어(학생):CPU,GPU
->비용

1지도학습:목표변수Y와 설명변수x의 관계를 수식화하여, 새로운 x가
들어올때, y를 예측또는 분류하는 모습

-회귀분석(Regressiom,y-연속형):주가예측/매출 예측/
-분류분석(Classification,y-범주형):스팸페일 판독/불량예측/고객이탈/이미지처리

2비지도학습:설명변수x간 유사성,연관성 등을 계산하여,
유사한데이터끼리 묶거나, 비슷한 데이터를 군집화하는 학습

-군집분석(Clustering)

-연관분석(Association Analysis):장바구니 분석/추천시스템

3강화학습:컴퓨터가 시뮬레이션을 통해 주어진 상황에 대해 (사용자 정의한 상황) 보상이 좋은 방향으로 학습(데이터가 없이 학습이 가능)
시뮬레이션 알고리즘을 만들어줘야함
게임AI/시뮬레이션

In [ ]:
import numpy as np
import pandas as pd
from google.colab import files
from google.colab import drive
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns #정적
import scipy.stats as stats
import plotly.express as px #동적

drive.mount('/content/drive')
trian = pd.read_csv("/content/drive/MyDrive/DSA/01_Data.csv")

mpl.rc('font',family='Malgun Gothic') #한글 맑은고딕

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
df1=pd.read_csv("/content/drive/MyDrive/DSA/01_Data.csv")

In [ ]:
df1.head(3)

Index  Member_ID Sales_Type Contract_Type Channel    Datetime  Term  \
0      1   66758234         렌탈          일반계약    영업방판  2019-05-06    60   
1      2   66755948         렌탈          교체계약    영업방판  2020-02-20    60   
2      3   66756657         렌탈          일반계약  홈쇼핑/방송  2019-02-28    60   

  Payment_Type Product_Type  Amount_Month Customer_Type   Age Address1  \
0          CMS        DES-1         96900            개인  42.0      경기도   
1         카드이체        DES-1        102900            개인  39.0      경기도   
2          CMS        DES-1         96900            개인  48.0      경기도   

  Address2 State  Overdue_count Overdue_Type Gender  Credit_Rank   Bank  
0      경기도  계약확정              0           없음     여자          9.0  새마을금고  
1      경기도  계약확정              0           없음     남자          2.0   현대카드  
2      경기도  계약확정              0           없음     여자          8.0   우리은행

#지도 학습 절차
계약 해지 여부 'state'항목을 목표변수Y로 계약 장보x를 이용해 예측(분류)하는 모델을 생성

1.데이터 헨들링:데이터를 불러와서 파생변수 생성/이상치 제거/변수 선택하는 작업
2.목표변수Y와 설명변수X를 선언
3.학습  데이터(Train set)와 검증데이터 (Test Set)로 분할
4.학습을 수행(Train Set)
-(특성공학/학습을 잘수행하기 위한 데이터 처리기법)

(결츠값처리/스케일링/인코딩/교차검증)

-학습 알고리즘에 의한 학습을 수행

5.평가(Evaluation)

-학습 능력 평가

-일반화 평가

6.실행

-Model을 사용 가능한 형태로 변환(파일/새로운 데이터 입력)

In [ ]:
#1,데이터 핸들링
#"state"항목을 "해약":1/"정상":0으로 데이터로 변환
df1['State'].value_counts()#목표변수

State
계약확정     50620
해약확정       622
기간만료        45
해약진행중       14
Name: count, dtype: int64

In [ ]:
#replace():특정 값을 다른 값으로 변환
df1['State'].replace({'계약확정':1,'해약확정':0,"기간만료":0,"해약진행중":1})

<ipython-input-43-d4d90ffb2dd2>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df1['State'].replace({'계약확정':1,'해약확정':0,"기간만료":0,"해약진행중":1})


0        1
1        1
2        1
3        1
4        0
        ..
51296    1
51297    1
51298    1
51299    1
51300    1
Name: State, Length: 51301, dtype: int64

In [ ]:
#replace():특정값을 다른 값으로 변환
df1['Target']=df1['State'].replace({'계약확정':1,'해약확정':0,"기간만료":0,"해약진행중":1})
df1['Target'].value_counts()

<ipython-input-44-03c6ba89b3fa>:2: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df1['Target']=df1['State'].replace({'계약확정':1,'해약확정':0,"기간만료":0,"해약진행중":1})


Target
1    50634
0      667
Name: count, dtype: int64

In [ ]:
#결측값제거
df2=df1.dropna()

In [ ]:
#2.목표변수Y와 설명변수X를 선언
Y=df2['Target']
df2

Index  Member_ID Sales_Type Contract_Type Channel    Datetime  Term  \
0          1   66758234         렌탈          일반계약    영업방판  2019-05-06    60   
1          2   66755948         렌탈          교체계약    영업방판  2020-02-20    60   
2          3   66756657         렌탈          일반계약  홈쇼핑/방송  2019-02-28    60   
3          4   66423450        멤버십        멤버십3유형     재계약  2019-05-13    12   
4          5   66423204        멤버십        멤버십3유형     재계약  2019-05-10    12   
...      ...        ...        ...           ...     ...         ...   ...   
51295  51297   66792261         렌탈          일반계약  홈쇼핑/방송  2019-03-11    60   
51297  51299   66799558         렌탈          일반계약   대형마트A  2019-04-01    60   
51298  51300   66799197         렌탈        프로모션계약    영업방판  2019-04-01    39   
51299  51301   66792778         렌탈          일반계약  홈쇼핑/방송  2020-02-06    60   
51300  51302   66799607         렌탈          일반계약  홈쇼핑/방송  2019-04-24    60   

      Payment_Type Product_Type  Amount_Month  ...   Age  Address1 Address2  \
0              CMS        DES-1         96900  ...  42.0       경기도      경기도   
1             카드이체        DES-1        102900  ...  39.0       경기도      경기도   
2              CMS        DES-1         96900  ...  48.0       경기도      경기도   
3              CMS        DES-1         66900  ...  39.0       경기도      경기도   
4              CMS        DES-1         66900  ...  60.0       경기도      경기도   
...            ...          ...           ...  ...   ...       ...      ...   
51295         카드이체        DES-1         96900  ...  45.0       경기도      경기도   
51297          CMS        DES-1         96900  ...  42.0       경기도      경기도   
51298         카드이체          ERA        120900  ...  65.0     서울특별시    서울특별시   
51299         카드이체        DES-1         96900  ...  54.0     서울특별시    서울특별시   
51300          CMS        DES-1         96900  ...  53.0     서울특별시    서울특별시   

      State Overdue_count  Overdue_Type Gender Credit_Rank    Bank Target  
0      계약확정             0            없음     여자         9.0   새마을금고      1  
1      계약확정             0            없음     남자         2.0    현대카드      1  
2      계약확정             0            없음     여자         8.0    우리은행      1  
3      계약확정             0            없음     남자         5.0  농협회원조합      1  
4         0            12            있음     남자         8.0  농협회원조합      0  
...     ...           ...           ...    ...         ...     ...    ...  
51295  계약확정             0            없음     여자         1.0    롯데카드      1  
51297  계약확정             0            없음     여자         8.0   새마을금고      1  
51298  계약확정             0            없음     여자         1.0    롯데카드      1  
51299  계약확정             0            없음     여자         2.0    롯데카드      1  
51300  계약확정             0            없음     여자         8.0    신한은행      1  

[40647 rows x 21 columns]

In [ ]:
X=df2[['Term',"Amount_Month","Age","Credit_Rank"]] #설명변수 x #2개이상일때 리스트2개

In [ ]:
#우회한 사람만 !pip--tristed-host pypi.org--tristed-host files.pythonhost.org install solkit-learn

In [ ]:
#3학습 데이터와 검증 데이터를 분할
from sklearn.model_selection import train_test_split
#from:라이브러리 내특정 클래스(특정동작을 수행하는 함수들이 모여있는 클래스를 호출)



In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, random_state=1234) #random_state=입력한 수가 동일한 값에 대해 랜덤하게 분할
 #unpinking기법 train_test_split():특정데이터를 무작위로 나누는 함수

In [ ]:
#4.학습을 수행
from sklearn.tree import DecisionTreeClassifier  # 분류학습을 도와줄 알고리즘


In [ ]:
model=DecisionTreeClassifier()
model.fit(X_trian,Y_train) #학습이 수행->Model 변수에 학습수식이 저장 정답지를 보여줌



DecisionTreeClassifier()

#새로운 데이터가 들어왔을 대 얼마나 잘되나 일반화
#학습능력평가


In [ ]:
#5.평가 수행
from sklearn.metrics import accuracy_score

In [ ]:
#생성한 모델에 x데이터를 넣어 모델이 게산한 예측값 확인
Y_train_pred=model.predict(X_trian) #이미 문제과정답을 본 학습지

In [ ]:
#학습능력평가
accuracy_score(Y_train,Y_train_pred)
#비교해서 정답지와 내가 만든 정답지  #문제만 보고 채점

0.9830736427751353

In [ ]:
#생성한 모델에 검증x데이터를 넣어 예측값 계산
Y_test_pred=model.predict(X_test)#새로운 데이터에서 학습

In [ ]:
#일반화 능력평가
accuracy_score(Y_test,Y_test_pred) #새로운 데이터에서 학습을 채점

0.9883881125762645

#정확도=정확하게 분류한 데이터수/전체데이터수
예)해약울 판별하는 분류모델 생성
1000개 학습 데이터->정상950/50
비율이 안맞는 것

정밀도=tp/tp+fp

재현율=tp/tp-fn

f1=2x정밀도x재현율/정밀도+재현율

In [ ]:
#분류모델 평가지표:계산함수
from sklearn.metrics import classification_report

In [ ]:
#학습능력평가
print(classification_report(Y_train,Y_train_pred))

              precision    recall  f1-score   support

           0       0.01      0.00      0.00       409
           1       0.99      1.00      0.99     30076

    accuracy                           0.98     30485
   macro avg       0.50      0.50      0.50     30485
weighted avg       0.97      0.98      0.98     30485



In [ ]:
#일반화 능력평가
print(classification_report(Y_test,Y_test_pred))

              precision    recall  f1-score   support

           0       0.55      0.19      0.28       122
           1       0.99      1.00      0.99     10040

    accuracy                           0.99     10162
   macro avg       0.77      0.59      0.64     10162
weighted avg       0.98      0.99      0.99     10162



In [ ]:
#6.실행/학습한 모델을 파일 형태로 저장

In [ ]:
import pickle

In [ ]:
pickle.dump(model,open("model.sav",'wb'))
#모델은 저장할 변수,머신러닝 모델은 보통 확장자 명을 sav파일로 저장함 wb는 컴퓨터

In [121]:
files.download('model.sav')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

AttributeError: 'NoneType' object has no attribute 'to_excel'